# Gegevensvoorbereiding

[Originele Notebook-bron van *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio door Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Verkennen van `DataFrame`-informatie

> **Leerdoel:** Aan het einde van deze subsectie moet je in staat zijn om algemene informatie over de gegevens die zijn opgeslagen in pandas DataFrames te vinden.

Zodra je je gegevens in pandas hebt geladen, zullen ze hoogstwaarschijnlijk in een `DataFrame` staan. Maar als de dataset in je `DataFrame` 60.000 rijen en 400 kolommen bevat, hoe begin je dan überhaupt een idee te krijgen van waar je mee werkt? Gelukkig biedt pandas enkele handige tools om snel algemene informatie over een `DataFrame` te bekijken, naast de eerste paar en laatste paar rijen.

Om deze functionaliteit te verkennen, importeren we de Python scikit-learn bibliotheek en gebruiken we een iconische dataset die elke datawetenschapper honderden keren heeft gezien: de *Iris*-dataset van de Britse bioloog Ronald Fisher, gebruikt in zijn paper uit 1936 "The use of multiple measurements in taxonomic problems":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
We hebben de Iris Dataset geladen in de variabele `iris_df`. Voordat we de data gaan verkennen, is het handig om te weten hoeveel datapunten we hebben en wat de totale omvang van de dataset is. Het is nuttig om een idee te krijgen van de hoeveelheid data waarmee we werken.


In [2]:
iris_df.shape

(150, 4)

We hebben hier te maken met 150 rijen en 4 kolommen aan data. Elke rij vertegenwoordigt één datapunt en elke kolom staat voor een enkele eigenschap die bij het data frame hoort. Kortom, er zijn 150 datapunten met elk 4 eigenschappen.

`shape` is hier een attribuut van het data frame en geen functie, daarom eindigt het niet met een paar haakjes.


### `DataFrame.columns`
Laten we nu de 4 kolommen van gegevens bekijken. Wat vertegenwoordigt elk van hen precies? De `columns`-eigenschap geeft ons de namen van de kolommen in de dataframe.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Zoals we kunnen zien, zijn er vier(4) kolommen. Het `columns`-attribuut vertelt ons de namen van de kolommen en verder eigenlijk niets. Dit attribuut wordt belangrijk wanneer we de kenmerken van een dataset willen identificeren.


### `DataFrame.info`
De hoeveelheid data (gegeven door het attribuut `shape`) en de namen van de kenmerken of kolommen (gegeven door het attribuut `columns`) vertellen ons iets over de dataset. Nu willen we dieper in de dataset duiken. De functie `DataFrame.info()` is hier erg handig voor.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


Van hieruit kunnen we een paar observaties maken:
1. Het datatype van elke kolom: In deze dataset worden alle gegevens opgeslagen als 64-bit floating-point getallen.
2. Aantal niet-nulwaarden: Het omgaan met nullwaarden is een belangrijke stap in de voorbereiding van gegevens. Dit zal later in het notebook worden behandeld.


### DataFrame.describe()
Stel dat we veel numerieke gegevens in onze dataset hebben. Univariate statistische berekeningen zoals het gemiddelde, de mediaan, kwartielen, enzovoort, kunnen afzonderlijk op elke kolom worden uitgevoerd. De functie `DataFrame.describe()` geeft ons een statistisch overzicht van de numerieke kolommen van een dataset.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

De bovenstaande output toont het totale aantal datapunten, gemiddelde, standaarddeviatie, minimum, onderste kwartiel (25%), mediaan (50%), bovenste kwartiel (75%) en de maximale waarde van elke kolom.


### `DataFrame.head`
Met alle bovenstaande functies en attributen hebben we een overzicht op hoog niveau van de dataset gekregen. We weten hoeveel datapunten er zijn, hoeveel kenmerken er zijn, het datatype van elk kenmerk en het aantal niet-nulwaarden voor elk kenmerk.

Nu is het tijd om naar de gegevens zelf te kijken. Laten we eens zien hoe de eerste paar rijen (de eerste paar datapunten) van onze `DataFrame` eruitzien:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Zoals we hier zien, zijn er vijf(5) vermeldingen van de dataset. Als we naar de index aan de linkerkant kijken, ontdekken we dat dit de eerste vijf rijen zijn.


### Oefening:

Uit het bovenstaande voorbeeld blijkt dat `DataFrame.head` standaard de eerste vijf rijen van een `DataFrame` retourneert. Kun je in de onderstaande codecel een manier vinden om meer dan vijf rijen weer te geven?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Een andere manier om naar de gegevens te kijken is vanaf het einde (in plaats van het begin). De tegenhanger van `DataFrame.head` is `DataFrame.tail`, die de laatste vijf rijen van een `DataFrame` retourneert:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

In de praktijk is het handig om eenvoudig de eerste paar rijen of de laatste paar rijen van een `DataFrame` te bekijken, vooral wanneer je op zoek bent naar uitschieters in geordende datasets.

Alle functies en attributen die hierboven zijn getoond met behulp van codevoorbeelden, helpen ons om een indruk te krijgen van de data.

> **Belangrijk punt:** Alleen al door te kijken naar de metadata over de informatie in een DataFrame of naar de eerste en laatste paar waarden ervan, kun je direct een idee krijgen van de grootte, vorm en inhoud van de data waarmee je werkt.


### Ontbrekende Gegevens
Laten we ons verdiepen in ontbrekende gegevens. Ontbrekende gegevens ontstaan wanneer er geen waarde is opgeslagen in sommige kolommen.

Laten we een voorbeeld nemen: stel dat iemand bewust is over zijn/haar gewicht en het veld voor gewicht niet invult in een enquête. Dan zal de waarde voor gewicht van die persoon ontbreken.

In de meeste gevallen komen ontbrekende waarden voor in datasets uit de echte wereld.

**Hoe Pandas omgaat met ontbrekende gegevens**

Pandas gaat op twee manieren om met ontbrekende waarden. De eerste manier heb je al eerder gezien in vorige secties: `NaN`, of Not a Number. Dit is eigenlijk een speciale waarde die deel uitmaakt van de IEEE floating-point specificatie en wordt alleen gebruikt om ontbrekende waarden van het type float aan te geven.

Voor ontbrekende waarden die geen floats zijn, gebruikt pandas het Python-object `None`. Hoewel het misschien verwarrend lijkt dat je twee verschillende soorten waarden tegenkomt die in essentie hetzelfde aangeven, zijn er goede programmatische redenen voor deze ontwerpkeuze. In de praktijk stelt deze aanpak pandas in staat om een goed compromis te bieden voor de overgrote meerderheid van de gevallen. Desondanks hebben zowel `None` als `NaN` beperkingen waar je rekening mee moet houden met betrekking tot hoe ze kunnen worden gebruikt.


### `None`: niet-vloeiende ontbrekende gegevens
Omdat `None` afkomstig is uit Python, kan het niet worden gebruikt in NumPy- en pandas-arrays die geen datatype `'object'` hebben. Onthoud dat NumPy-arrays (en de datastructuren in pandas) slechts één type gegevens kunnen bevatten. Dit is wat hen hun enorme kracht geeft voor grootschalige data- en rekentaken, maar het beperkt ook hun flexibiliteit. Dergelijke arrays moeten worden opgewaardeerd naar de "laagste gemene deler," het datatype dat alles in de array kan omvatten. Wanneer `None` in de array voorkomt, betekent dit dat je werkt met Python-objecten.

Om dit in de praktijk te zien, bekijk het volgende voorbeeld van een array (let op de `dtype` ervan):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

De realiteit van opgehoogde datatypes brengt twee bijwerkingen met zich mee. Ten eerste worden bewerkingen uitgevoerd op het niveau van geïnterpreteerde Python-code in plaats van gecompileerde NumPy-code. In essentie betekent dit dat alle bewerkingen met `Series` of `DataFrames` waarin `None` voorkomt, langzamer zullen zijn. Hoewel je deze prestatievermindering waarschijnlijk niet zult opmerken, kan het bij grote datasets een probleem worden.

De tweede bijwerking vloeit voort uit de eerste. Omdat `None` `Series` of `DataFrames` feitelijk terugbrengt naar de wereld van standaard Python, zal het gebruik van NumPy/pandas-aggregaties zoals `sum()` of `min()` op arrays die een ``None``-waarde bevatten, doorgaans een fout veroorzaken:


In [10]:
example1.sum()

TypeError: ignored

**Belangrijkste punt**: Optelling (en andere bewerkingen) tussen gehele getallen en `None`-waarden is ongedefinieerd, wat kan beperken wat je kunt doen met datasets die deze bevatten.


### `NaN`: ontbrekende float-waarden

In tegenstelling tot `None` ondersteunt NumPy (en dus ook pandas) `NaN` voor zijn snelle, gevectoriseerde bewerkingen en ufuncs. Het nadeel is dat elke rekenkundige bewerking met `NaN` altijd resulteert in `NaN`. Bijvoorbeeld:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

Het goede nieuws: aggregaties uitgevoerd op arrays met `NaN` erin veroorzaken geen fouten. Het slechte nieuws: de resultaten zijn niet uniform bruikbaar:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Oefening:


In [11]:
# What happens if you add np.nan and None together?


Onthoud: `NaN` is alleen voor ontbrekende waarden met drijvende komma; er is geen `NaN`-equivalent voor gehele getallen, strings of Booleans.


### `NaN` en `None`: nullwaarden in pandas

Hoewel `NaN` en `None` zich enigszins anders kunnen gedragen, is pandas toch ontworpen om ze door elkaar te gebruiken. Om te begrijpen wat we bedoelen, bekijk een `Series` van gehele getallen:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Oefening:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


Bij het opwaarderen van gegevenstypen om gegevenshomogeniteit te bereiken in `Series` en `DataFrame`s, zal pandas zonder problemen ontbrekende waarden wisselen tussen `None` en `NaN`. Door deze ontwerpkeuze kan het nuttig zijn om `None` en `NaN` te beschouwen als twee verschillende varianten van "null" in pandas. Sterker nog, sommige van de kernmethoden die je zult gebruiken om met ontbrekende waarden in pandas om te gaan, weerspiegelen dit idee in hun namen:

- `isnull()`: Genereert een Booleaanse maskering die ontbrekende waarden aangeeft
- `notnull()`: Tegenovergestelde van `isnull()`
- `dropna()`: Geeft een gefilterde versie van de gegevens terug
- `fillna()`: Geeft een kopie van de gegevens terug waarbij ontbrekende waarden zijn ingevuld of geïmputeerd

Dit zijn belangrijke methoden om onder de knie te krijgen en vertrouwd mee te raken, dus laten we ze elk wat uitgebreider bespreken.


### Nullwaarden detecteren

Nu we het belang van ontbrekende waarden hebben begrepen, moeten we ze in onze dataset opsporen voordat we ermee aan de slag gaan. Zowel `isnull()` als `notnull()` zijn je belangrijkste methoden om nullwaarden te detecteren. Beide geven Booleaanse maskers over je data terug.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Kijk goed naar de output. Verrast iets je? Hoewel `0` een wiskundige nul is, is het toch een volwaardig geheel getal en pandas behandelt het ook zo. `''` is iets subtieler. Hoewel we het in Sectie 1 gebruikten om een lege tekenreekswaarde weer te geven, is het toch een tekenreeksobject en geen representatie van null volgens pandas.

Laten we dit nu omdraaien en deze methoden gebruiken op een manier zoals je ze in de praktijk zult toepassen. Je kunt Booleaanse maskers direct gebruiken als een ``Series``- of ``DataFrame``-index, wat handig kan zijn wanneer je wilt werken met geïsoleerde ontbrekende (of aanwezige) waarden.

Als we het totale aantal ontbrekende waarden willen weten, kunnen we gewoon een som nemen over het masker dat door de `isnull()`-methode wordt geproduceerd.


In [19]:
example3.isnull().sum()

2

### Oefening:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Belangrijkste inzicht**: Zowel de methoden `isnull()` als `notnull()` leveren vergelijkbare resultaten op wanneer je ze gebruikt in DataFrames: ze tonen de resultaten en de index van die resultaten, wat je enorm zal helpen bij het werken met je data.


### Omgaan met ontbrekende gegevens

> **Leerdoel:** Aan het einde van deze subsectie weet je hoe en wanneer je null-waarden in DataFrames kunt vervangen of verwijderen.

Machine Learning-modellen kunnen niet zelf omgaan met ontbrekende gegevens. Daarom moeten we, voordat we de gegevens aan het model doorgeven, deze ontbrekende waarden aanpakken.

Hoe ontbrekende gegevens worden behandeld, brengt subtiele afwegingen met zich mee en kan invloed hebben op je uiteindelijke analyse en resultaten in de praktijk.

Er zijn voornamelijk twee manieren om met ontbrekende gegevens om te gaan:

1.   De rij met de ontbrekende waarde verwijderen
2.   De ontbrekende waarde vervangen door een andere waarde

We zullen beide methoden en hun voor- en nadelen uitgebreid bespreken.


### Nullwaarden verwijderen

De hoeveelheid data die we aan ons model doorgeven, heeft een directe invloed op de prestaties ervan. Nullwaarden verwijderen betekent dat we het aantal datapunten verminderen en daarmee de grootte van de dataset verkleinen. Het is daarom aan te raden om rijen met nullwaarden te verwijderen wanneer de dataset behoorlijk groot is.

Een andere situatie kan zijn dat een bepaalde rij of kolom veel ontbrekende waarden heeft. In dat geval kunnen ze worden verwijderd, omdat ze weinig waarde toevoegen aan onze analyse, aangezien de meeste gegevens voor die rij/kolom ontbreken.

Naast het identificeren van ontbrekende waarden biedt pandas een handige manier om nullwaarden te verwijderen uit `Series` en `DataFrame`s. Om dit in de praktijk te zien, laten we teruggaan naar `example3`. De functie `DataFrame.dropna()` helpt bij het verwijderen van rijen met nullwaarden.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Merk op dat dit eruit zou moeten zien als je uitvoer van `example3[example3.notnull()]`. Het verschil hier is dat, in plaats van alleen te indexeren op de gemaskeerde waarden, `dropna` die ontbrekende waarden uit de `Series` `example3` heeft verwijderd.

Omdat DataFrames twee dimensies hebben, bieden ze meer opties om gegevens te verwijderen.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Heb je gemerkt dat pandas twee van de kolommen heeft omgezet naar floats om de `NaN`s te accommoderen?)

Je kunt geen enkele waarde uit een `DataFrame` verwijderen, dus je moet volledige rijen of kolommen verwijderen. Afhankelijk van wat je aan het doen bent, wil je misschien het een of het ander doen, en daarom biedt pandas opties voor beide. Omdat in datawetenschap kolommen meestal variabelen vertegenwoordigen en rijen observaties, is de kans groter dat je rijen met gegevens verwijdert; de standaardinstelling voor `dropna()` is om alle rijen te verwijderen die null-waarden bevatten:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Indien nodig kun je NA-waarden uit kolommen verwijderen. Gebruik `axis=1` om dit te doen:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Merk op dat dit veel gegevens kan verwijderen die je misschien wilt behouden, vooral bij kleinere datasets. Wat als je alleen rijen of kolommen wilt verwijderen die meerdere of zelfs alle nullwaarden bevatten? Je kunt deze instellingen specificeren in `dropna` met de parameters `how` en `thresh`.

Standaard is `how='any'` (als je dit zelf wilt controleren of wilt zien welke andere parameters de methode heeft, voer dan `example4.dropna?` uit in een codecel). Je kunt ook `how='all'` specificeren om alleen rijen of kolommen te verwijderen die volledig uit nullwaarden bestaan. Laten we ons voorbeeld `DataFrame` uitbreiden om dit in actie te zien in de volgende oefening.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Belangrijkste punten:  
1. Het verwijderen van null-waarden is alleen een goed idee als de dataset groot genoeg is.  
2. Volledige rijen of kolommen kunnen worden verwijderd als het merendeel van de gegevens ontbreekt.  
3. De methode `DataFrame.dropna(axis=)` helpt bij het verwijderen van null-waarden. Het argument `axis` geeft aan of rijen of kolommen moeten worden verwijderd.  
4. Het argument `how` kan ook worden gebruikt. Standaard is het ingesteld op `any`. Dit betekent dat alleen die rijen/kolommen worden verwijderd die een null-waarde bevatten. Het kan worden ingesteld op `all` om aan te geven dat we alleen die rijen/kolommen verwijderen waar alle waarden null zijn.  


### Oefening:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


De `thresh`-parameter geeft je meer gedetailleerde controle: je stelt het aantal *niet-nulle* waarden in dat een rij of kolom moet hebben om behouden te blijven:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

Hier zijn de eerste en laatste rij verwijderd, omdat ze slechts twee niet-nulwaarden bevatten.


### Nullwaarden invullen

Het kan soms logisch zijn om ontbrekende waarden in te vullen met waarden die geldig zouden kunnen zijn. Er zijn een paar technieken om nullwaarden in te vullen. De eerste is het gebruik van Domeinkennis (kennis van het onderwerp waarop de dataset is gebaseerd) om op een of andere manier de ontbrekende waarden te benaderen.

Je kunt `isnull` gebruiken om dit direct te doen, maar dat kan tijdrovend zijn, vooral als je veel waarden moet invullen. Omdat dit een veelvoorkomende taak is in data science, biedt pandas `fillna`, waarmee je een kopie van de `Series` of `DataFrame` krijgt waarin de ontbrekende waarden zijn vervangen door een waarde naar keuze. Laten we een ander voorbeeld van een `Series` maken om te zien hoe dit in de praktijk werkt.


### Categorische gegevens (Niet-numeriek)
Laten we eerst niet-numerieke gegevens bekijken. In datasets hebben we kolommen met categorische gegevens, zoals geslacht, waar of onwaar, enz.

In de meeste gevallen vervangen we ontbrekende waarden door de `modus` van de kolom. Stel dat we 100 datapunten hebben, waarvan 90 "waar" hebben aangegeven, 8 "onwaar" en 2 niets hebben ingevuld. Dan kunnen we de 2 invullen met "waar", rekening houdend met de hele kolom.

Ook hier kunnen we domeinkennis gebruiken. Laten we een voorbeeld bekijken van invullen met de modus.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Laten we eerst de modus vinden voordat we de `None`-waarde vullen met de modus.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

Dus, we zullen None vervangen door True


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Zoals we kunnen zien, is de null-waarde vervangen. Het is vanzelfsprekend dat we alles hadden kunnen schrijven in plaats van `'True'` en het zou zijn vervangen.


### Numerieke Gegevens
Nu gaan we verder met numerieke gegevens. Hier zijn twee veelgebruikte methoden om ontbrekende waarden te vervangen:

1. Vervangen door de mediaan van de rij
2. Vervangen door het gemiddelde van de rij

We vervangen door de mediaan in het geval van scheve gegevens met uitschieters. Dit komt omdat de mediaan robuust is tegen uitschieters.

Wanneer de gegevens genormaliseerd zijn, kunnen we het gemiddelde gebruiken, omdat in dat geval het gemiddelde en de mediaan dicht bij elkaar liggen.

Laten we eerst een kolom nemen die normaal verdeeld is en de ontbrekende waarde invullen met het gemiddelde van de kolom.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

Het gemiddelde van de kolom is


In [33]:
np.mean(fill_with_mean[0])

0.0

Vullen met gemiddelde


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Zoals we kunnen zien, is de ontbrekende waarde vervangen door het gemiddelde.


Laten we nu een andere dataframe proberen, en deze keer zullen we de None-waarden vervangen door de mediaan van de kolom.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

De mediaan van de tweede kolom is


In [36]:
fill_with_median[1].median()

4.0

Vullen met mediaan


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Zoals we kunnen zien, is de NaN-waarde vervangen door de mediaan van de kolom


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Je kunt alle lege invoervelden vullen met een enkele waarde, zoals `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Belangrijke punten:
1. Het invullen van ontbrekende waarden moet worden gedaan wanneer er weinig gegevens zijn of wanneer er een strategie is om de ontbrekende gegevens in te vullen.
2. Domeinkennis kan worden gebruikt om ontbrekende waarden te benaderen en in te vullen.
3. Voor categorische gegevens worden ontbrekende waarden meestal vervangen door de modus van de kolom.
4. Voor numerieke gegevens worden ontbrekende waarden meestal ingevuld met het gemiddelde (voor genormaliseerde datasets) of de mediaan van de kolommen.


### Oefening:


In [40]:
# What happens if you try to fill null values with a string, like ''?


Je kunt **forward-fill** nullwaarden, wat betekent dat je de laatste geldige waarde gebruikt om een nullwaarde in te vullen:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Je kunt ook **back-fill** gebruiken om de volgende geldige waarde achterwaarts te propageren om een null te vullen:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Zoals je misschien kunt raden, werkt dit hetzelfde met DataFrames, maar je kunt ook een `as` specificeren langs welke null-waarden moeten worden ingevuld:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Merk op dat wanneer een eerdere waarde niet beschikbaar is voor forward-filling, de null-waarde blijft.


### Oefening:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Je kunt creatief zijn met hoe je `fillna` gebruikt. Laten we bijvoorbeeld nog eens naar `example4` kijken, maar deze keer vullen we de ontbrekende waarden in met het gemiddelde van alle waarden in de `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Merk op dat kolom 3 nog steeds geen waarde heeft: de standaardrichting is om waarden rij per rij in te vullen.

> **Belangrijk punt:** Er zijn meerdere manieren om om te gaan met ontbrekende waarden in je datasets. De specifieke strategie die je kiest (verwijderen, vervangen, of zelfs hoe je ze vervangt) moet worden bepaald door de specifieke kenmerken van die data. Hoe meer je met datasets werkt en interactie hebt, hoe beter je gevoel wordt voor het omgaan met ontbrekende waarden.


### Coderen van Categorische Gegevens

Machine learning-modellen werken alleen met cijfers en elke vorm van numerieke data. Ze kunnen geen onderscheid maken tussen een Ja en een Nee, maar wel tussen 0 en 1. Dus, nadat we de ontbrekende waarden hebben ingevuld, moeten we de categorische gegevens coderen naar een numerieke vorm zodat het model ze kan begrijpen.

Coderen kan op twee manieren worden gedaan. We zullen deze hierna bespreken.


**LABELCODERING**

Labelcodering houdt in dat elke categorie wordt omgezet in een nummer. Stel bijvoorbeeld dat we een dataset hebben van vliegtuigpassagiers en er is een kolom met hun klasse, zoals ['business class', 'economy class', 'first class']. Als hier labelcodering op wordt toegepast, wordt dit omgezet naar [0, 1, 2]. Laten we een voorbeeld bekijken via code. Omdat we `scikit-learn` in de komende notebooks gaan leren, gebruiken we het hier niet.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Om label encoding uit te voeren op de 1e kolom, moeten we eerst een mapping beschrijven van elke klasse naar een nummer, voordat we vervangen.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Zoals we kunnen zien, komt de output overeen met wat we hadden verwacht. Dus, wanneer gebruiken we label encoding? Label encoding wordt gebruikt in een of beide van de volgende gevallen:
1. Wanneer het aantal categorieën groot is
2. Wanneer de categorieën een volgorde hebben.


**ONE HOT ENCODING**

Een andere vorm van codering is One Hot Encoding. Bij deze vorm van codering wordt elke categorie van de kolom toegevoegd als een aparte kolom, en elk datapunt krijgt een 0 of een 1 afhankelijk van of het die categorie bevat. Dus, als er n verschillende categorieën zijn, worden er n kolommen toegevoegd aan de dataframe.

Bijvoorbeeld, laten we hetzelfde voorbeeld van vliegtuigklassen nemen. De categorieën waren: ['business class', 'economy class', 'first class']. Als we One Hot Encoding toepassen, worden de volgende drie kolommen toegevoegd aan de dataset: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Laten we one-hot encoding uitvoeren op de eerste kolom


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Elke one-hot-gecodeerde kolom bevat 0 of 1, wat aangeeft of die categorie bestaat voor dat datapunt.


Wanneer gebruiken we one-hot encoding? One-hot encoding wordt gebruikt in een of beide van de volgende gevallen:

1. Wanneer het aantal categorieën en de grootte van de dataset klein zijn.
2. Wanneer de categorieën geen specifieke volgorde hebben.


> Belangrijkste punten:
1. Codering wordt gebruikt om niet-numerieke gegevens om te zetten in numerieke gegevens.
2. Er zijn twee soorten codering: Labelcodering en One Hot-codering, die beide kunnen worden uitgevoerd afhankelijk van de eisen van de dataset.


## Duplicaatgegevens verwijderen

> **Leerdoel:** Aan het einde van deze subsectie moet je in staat zijn om duplicaatwaarden in DataFrames te identificeren en te verwijderen.

Naast ontbrekende gegevens kom je vaak duplicaatgegevens tegen in datasets uit de praktijk. Gelukkig biedt pandas een eenvoudige manier om duplicaatvermeldingen te detecteren en te verwijderen.


### Duplicaten identificeren: `duplicated`

Je kunt eenvoudig dubbele waarden opsporen met de `duplicated`-methode in pandas, die een Booleaanse maskering retourneert om aan te geven of een invoer in een `DataFrame` een duplicaat is van een eerdere. Laten we een ander voorbeeld `DataFrame` maken om dit in actie te zien.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Duplicaten verwijderen: `drop_duplicates`
`drop_duplicates` geeft eenvoudigweg een kopie van de gegevens terug waarvoor alle `duplicated` waarden `False` zijn:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Zowel `duplicated` als `drop_duplicates` beschouwen standaard alle kolommen, maar je kunt specificeren dat ze alleen een subset van kolommen in je `DataFrame` onderzoeken:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **Belangrijk:** Het verwijderen van dubbele gegevens is een essentieel onderdeel van bijna elk data-scienceproject. Dubbele gegevens kunnen de resultaten van je analyses veranderen en je onnauwkeurige resultaten geven!


## Kwaliteitscontroles van gegevens uit de praktijk

> **Leerdoel:** Aan het einde van deze sectie moet je in staat zijn om veelvoorkomende kwaliteitsproblemen in gegevens uit de praktijk te herkennen en te corrigeren, zoals inconsistente categoriewaarden, abnormale numerieke waarden (uitbijters) en dubbele entiteiten met variaties.

Hoewel ontbrekende waarden en exacte duplicaten veelvoorkomende problemen zijn, bevatten datasets uit de praktijk vaak subtielere problemen:

1. **Inconsistente categoriewaarden**: Dezelfde categorie anders gespeld (bijv. "USA", "U.S.A", "United States")
2. **Abnormale numerieke waarden**: Extreme uitbijters die wijzen op invoerfouten (bijv. leeftijd = 999)
3. **Bijna-duplicaten**: Records die dezelfde entiteit vertegenwoordigen met kleine variaties

Laten we technieken verkennen om deze problemen te detecteren en aan te pakken.


### Een voorbeeld van een "vuile" dataset maken

Laten we eerst een voorbeelddataset maken die de soorten problemen bevat die we vaak tegenkomen in echte data:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. Detecteren van inconsistente categorische waarden

Let op dat de `country`-kolom meerdere representaties heeft voor dezelfde landen. Laten we deze inconsistenties identificeren:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### Standaardiseren van Categorische Waarden

We kunnen een mapping maken om deze waarden te standaardiseren. Een eenvoudige aanpak is om ze om te zetten naar kleine letters en een mapping dictionary te maken:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**Alternatief: Gebruik van fuzzy matching**

Voor meer complexe gevallen kunnen we fuzzy string matching gebruiken met de `rapidfuzz`-bibliotheek om automatisch vergelijkbare strings te detecteren:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. Detecteren van Abnormale Numerieke Waarden (Uitbijters)

Als we naar de `age`-kolom kijken, zien we enkele verdachte waarden zoals 199 en -5. Laten we statistische methoden gebruiken om deze uitbijters te detecteren.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### Gebruik van de IQR (Interkwartielafstand) Methode

De IQR-methode is een robuuste statistische techniek voor het detecteren van uitschieters die minder gevoelig is voor extreme waarden:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Gebruik van de Z-Score Methode

De Z-score methode identificeert uitschieters op basis van standaarddeviaties vanaf het gemiddelde:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### Omgaan met uitschieters

Zodra uitschieters zijn gedetecteerd, kunnen ze op verschillende manieren worden behandeld:
1. **Verwijderen**: Verwijder rijen met uitschieters (als het fouten zijn)
2. **Afschermen**: Vervang door grenswaarden
3. **Vervangen door NaN**: Behandel als ontbrekende gegevens en gebruik imputatietechnieken
4. **Behouden**: Als het legitieme extreme waarden zijn


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. Detecteren van bijna-duplicaat rijen

Merk op dat onze dataset meerdere vermeldingen heeft voor "John Smith" met licht verschillende waarden. Laten we potentiële duplicaten identificeren op basis van naamgelijkenis.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### Nabijgelegen duplicaten vinden met fuzzy matching

Voor meer geavanceerde duplicaatdetectie kunnen we fuzzy matching gebruiken om vergelijkbare namen te vinden:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### Omgaan met duplicaten

Zodra ze zijn geïdentificeerd, moet je beslissen hoe je met duplicaten omgaat:
1. **Behoud de eerste voorkoming**: Gebruik `drop_duplicates(keep='first')`
2. **Behoud de laatste voorkoming**: Gebruik `drop_duplicates(keep='last')`
3. **Informatie samenvoegen**: Combineer informatie uit dubbele rijen
4. **Handmatige controle**: Markeer voor menselijke beoordeling


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### Samenvatting: Volledige Gegevensopschoonpipeline

Laten we alles samenvoegen tot een uitgebreide opschoonpipeline:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 Uitdagingsoefening

Nu is het jouw beurt! Hieronder staat een nieuwe rij met gegevens die meerdere kwaliteitsproblemen bevat. Kun je:

1. Alle problemen in deze rij identificeren
2. Code schrijven om elk probleem op te lossen
3. De schoongemaakte rij toevoegen aan de dataset

Hier zijn de problematische gegevens:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### Belangrijke punten

1. **Inconsistente categorieën** komen vaak voor in echte datasets. Controleer altijd de unieke waarden en standaardiseer ze met behulp van mappings of fuzzy matching.

2. **Uitschieters** kunnen je analyse aanzienlijk beïnvloeden. Gebruik domeinkennis in combinatie met statistische methoden (IQR, Z-score) om ze te detecteren.

3. **Bijna-duplicaten** zijn moeilijker te herkennen dan exacte duplicaten. Overweeg het gebruik van fuzzy matching en het normaliseren van gegevens (kleine letters, spaties verwijderen) om ze te identificeren.

4. **Gegevens opschonen is iteratief**. Je moet mogelijk meerdere technieken toepassen en de resultaten beoordelen voordat je je opgeschoonde dataset definitief maakt.

5. **Documenteer je beslissingen**. Houd bij welke schoonmaakstappen je hebt toegepast en waarom, omdat dit belangrijk is voor reproduceerbaarheid en transparantie.

> **Beste praktijk:** Bewaar altijd een kopie van je originele "vuile" data. Overschrijf nooit je bronbestanden - maak opgeschoonde versies met duidelijke naamconventies zoals `data_cleaned.csv`.



---

**Disclaimer**:  
Dit document is vertaald met behulp van de AI-vertalingsservice [Co-op Translator](https://github.com/Azure/co-op-translator). Hoewel we streven naar nauwkeurigheid, dient u zich ervan bewust te zijn dat geautomatiseerde vertalingen fouten of onnauwkeurigheden kunnen bevatten. Het originele document in de oorspronkelijke taal moet worden beschouwd als de gezaghebbende bron. Voor cruciale informatie wordt professionele menselijke vertaling aanbevolen. Wij zijn niet aansprakelijk voor misverstanden of verkeerde interpretaties die voortvloeien uit het gebruik van deze vertaling.
